### PR TITLE
Add Windows Support

### LINKED ISSUE #1504: Add Windows support
Some users have requested Windows support for Kuzu, such as #1311 and #1313. The steps required to support Windows are as follows:

- [ ] Fix compilation errors for the core library and test cases with MSVC.
- [ ] Run the test cases, fix any testing error.
- [ ] Build a dll version for `libkuzu` and test it with the MSVC toolchain.
- [ ] Figure out how to build the Python wheel for Windows.
- [ ] Replace [linenoise](https://github.com/antirez/linenoise) with another Windows-compatible readline implementation for the shell.
- [ ] Setup a Windows-based CI runner to ensure the compatibility with Windows for the PRs afterwards.

I worked on Windows support on branch https://github.com/kuzudb/kuzu/tree/windows by attempting fixing all the compilation errors and replace system API calls. However, there is one compilation error regarding template that I did not know how to fix (in `src\storage\storage_structure\lists\lists_update_store.cpp`). The branch is stale now since a lot of new features were added in version 0.0.3. I no longer have the detailed output from MSVC, but it can be reproduced by checking out the branch and run the build process in Windows PowerShell.

### PR DESCRIPTION
**Pull request by benjaminwinger**
A few tests seem to have regressed with the most recent changes from master. I haven't gotten around to looking into those yet, but I thought I should get this PR open for review in the meantime.

Mostly fixes #1504 (note omissions below) and supersedes #1544.

### Notable changes

I added a `KUZU_` prefix to the type enum in the C API. Unfortunately, windows headers pollute the import namespace making some of the more common names (e.g. `BOOL`) ambiguous.

I removed the `VectorOperations` class and moved `{Unary,Binary,Ternary,Const}ExecFunction` into the enclosing `kuzu::function` namespace (they could get their own child namespace if that's preferable). MSVC doesn't appear to allow function pointers to templated member functions (see [this minimal example](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAGykrgAyeAyYAHI%2BAEaYxCAAHKQADqgKhE4MHt6%2BASlpGQKh4VEssfFJdpgOmUIETMQE2T5%2BgVU1AnUNBMWRMXGJtvWNzbltwz1hfWUDCQCUtqhexMjsHAD06wDUACoAnsmYW3tLxFtoWFsIcZikWyRbtKhM6FuGW5iqrMn0AHQmGgAguYAMxhZDeS4mEFuKheBgdZi0aHYAHAoEQpgKBS7TAKAhbEwAdisQOSXmitDwyBAaK29K2%2BKYjmQjII6BAIDhCMcAmhbgAbqg8K8IGECHMUVtomF0AAxeGIiBzQkkukMjXETAEZYMQlmfw7PEETkEY0AfW5iP54pR0NJgI1xIAImjksQ8ALmex1fSzSwft7%2BQQDoxWEdAeadgBNZSokGooEapksrZCkVbM34y2K3kMCCRmNxtNiLyYFXEh0ap1JhkutH1kEOtHp15ZgjK1VVhlMLxELZWwkg524/HK34yhjy3OZZX230DxUQLgaSVNhtE50cBa0TgAVl4fg4WlIqE4bms1kZp1W%2BpBPFIJuP24WAGsQGY9/pOJJD5peGeHC8AoIAaI%2B/4LHAsBIGgAZ0HE5CULByTwfEArIMkyTmgKXAAJzmgY7bmqo/jSFgArUpgABqeCYAA7gA8ocR4PjQtBmsQIEQNE/6kJODR7JwD78cQewMdE2jVE%2BD6wWwggMQwtCCc%2BpBYNEXjAG4Yi0CB3C8FgLCGMA4gqfgWo1AKeK8Z81R9msD7ipgu4qVS0TEAJHhYLxBAeiwQnbnwBjAAoNH0UxjD%2BTIggiGI7BSFF8hKGovG6Fw%2BhGSgl6WPoeDRCBkALKgyR5rpAGWcQHpYPlyq2E5UmZC4U6jH4khfiEUylOUIAgvk6R5s1ICtb1hQML0nUDD17R5l0IyeC0g1flNtQTGN/TxJNEwDUNTKNKtMzrQsCg3nFO77n%2BKmAVsJGSFsLAKOhaZ4b8hHGlsEC4IQDyglwcy8E%2BWhzG%2BH5fs5v6kEeJ6AcBoHgc%2BkEwIgIBLAQ5IEIhEDIahEThpwqgJP4AC0pFbMAyCslwIK/A%2BmD4EQlV6Pw0WiOI8WM4lKjqCpqWkHR7nJP5p0cAe4O8YBDF9qj9xUFdxN3Q9OG4c93r4m9HhwfQZzfb9sMA0DZiSL8uFG8bJsm4EoPnZDnDQ2B/0BVBiOYxr6NOwMwAU2l5GUaFjHMZFbEcVxPEqSJynCWEAniZJDiRbJjAEApSm8WpGlabQOmRQZRkmSeZn1ZZuknjZyB2ZFjnOSernuaJnlrCePl4H5ekLFQQUhbRvsRXpCUxSz0hs4oHMpR%2B6VGJlljWDleXwIVxWZKVp7lZVVkFbV9h5o17hzbkaXtSUa16KkfWZANaVHyNe1dWlS2dJt29%2BNfdWIjNkz7/teg7U098fytHUHz9ixlirAkILYWEMAKcCuvjImN1Sbk0plwN6H06Z3h%2Bn9CCesQY/kthAoCtgYZ20Bt%2BDgZgcGnmtjrF8pByrpGcJIIAA%3D%3D)).

### Omissions
- The shell is not being built on windows as linenoise isn't yet working (it looks like the simplest option may be to try to get the current version of linenoise being used working with windows via some patches which exist for older versions of linenoise).
- The python build is working, but a CI job for creating wheels has not been set up.
- Bulding Debug or RelWithDebInfo on windows is tricky due to how the arrow build is set up and the fact that debug and release builds can't be linked together on windows, so it needs to match the main build. I've added an `ARROW_INSTALL` option which can point to the installed arrow prefix and can be used to switch between versions of arrow manually. Setting something up in the Makefile to do it automatically should be possible, but would probably be messy.

Also note that performance is rather poor. It seems to largely be related to file i/o.

### CODE REVIEW

File: src/include/function/boolean/vector_boolean_operations.h [hunk 1]
```diff
@@ -6,7 +6,39 @@
 namespace kuzu {
 namespace function {
 
-class VectorBooleanOperations : public VectorOperations {
+template<typename FUNC>
```
Comment (andyfengHKU) [inline src/include/function/boolean/vector_boolean_operations.h:9]: Can u try to put these back into `VectorBooleanOperations`. This probably won't cause compilation problem in MSVC.

File: src/include/function/comparison/vector_comparison_operations.h [hunk 2]
```diff
@@ -164,10 +164,10 @@ class VectorComparisonOperations : public VectorOperations {
         case common::LogicalTypeID::TIMESTAMP: {
             switch (rightTypeID) {
             case common::LogicalTypeID::DATE: {
-                return BinarySelectFunction<common::timestamp_t, common::date_t, FUNC>;
+                return &BinarySelectFunction<common::timestamp_t, common::date_t, FUNC>;
             }
             case common::LogicalTypeID::TIMESTAMP: {
-                return BinarySelectFunction<common::timestamp_t, common::timestamp_t, FUNC>;
+                return &kuzu::function::BinarySelectFunction<common::timestamp_t, common::timestamp_t, FUNC>;
```
Comment (andyfengHKU) [inline src/include/function/comparison/vector_comparison_operations.h:170]: remove kuzu::function

File: src/include/function/list/vector_list_operations.h [hunk 3]
```diff
@@ -5,26 +5,33 @@
...
-    template<typename LEFT_TYPE, typename RIGHT_TYPE, typename RESULT_TYPE, typename FUNC>
-    static void BinaryListExecFunction(
-        const std::vector<std::shared_ptr<common::ValueVector>>& params,
-        common::ValueVector& result) {
-        assert(params.size() == 2);
-        BinaryOperationExecutor::executeList<LEFT_TYPE, RIGHT_TYPE, RESULT_TYPE, FUNC>(
-            *params[0], *params[1], result);
-    }
-
+template<typename A_TYPE, typename B_TYPE, typename C_TYPE, typename RESULT_TYPE, typename FUNC>
```
Comment (andyfengHKU) [inline src/include/function/list/vector_list_operations.h:8]: ditto. try put back into `VectorListOperations`

File: src/include/function/null/vector_null_operations.h [hunk 4]
```diff
@@ -6,7 +6,23 @@
 namespace kuzu {
 namespace function {
 
-class VectorNullOperations : public VectorOperations {
+template<typename FUNC>
```
Comment (andyfengHKU) [inline src/include/function/null/vector_null_operations.h:9]: ditto try put back into `VectorNullOperations`

File: test/demo_db/demo_db_test.cpp [hunk 5]
```diff
@@ -97,15 +97,6 @@ TEST_F(DemoDBTest, DeleteRelTest) {
     ASSERT_EQ(TestHelper::convertResultToString(*result), groundTruth);
 }
 
-TEST_F(DemoDBTest, DeleteWithExceptionTest) {
```
Comment (andyfengHKU) [inline test/demo_db/demo_db_test.cpp:100]: Why is this removed?

File: src/main/connection.cpp [hunk 6]
```diff
@@ -212,17 +212,27 @@ std::unique_ptr<PreparedStatement> Connection::prepareNoLock(
 std::string Connection::getNodeTableNames() {
     lock_t lck{mtx};
     std::string result = "Node tables: \n";
+    std::vector<std::string> nodeTableNames;
     for (auto& tableIDSchema : database->catalog->getReadOnlyVersion()->getNodeTableSchemas()) {
-        result += "\t" + tableIDSchema.second->tableName + "\n";
+        nodeTableNames.push_back(tableIDSchema.second->tableName);
+    }
+    std::sort(nodeTableNames.begin(), nodeTableNames.end());
```
Comment (andyfengHKU) [inline src/main/connection.cpp:219]: do we need to sort. We should print in the same order as defined in catalog

File: src/main/connection.cpp [hunk 7]
```diff
@@ -212,17 +212,27 @@ std::unique_ptr<PreparedStatement> Connection::prepareNoLock(
...
 
 std::string Connection::getRelTableNames() {
     lock_t lck{mtx};
     std::string result = "Rel tables: \n";
+    std::vector<std::string> relTableNames;
     for (auto& tableIDSchema : database->catalog->getReadOnlyVersion()->getRelTableSchemas()) {
-        result += "\t" + tableIDSchema.second->tableName + "\n";
+        relTableNames.push_back(tableIDSchema.second->tableName);
+    }
+    std::sort(relTableNames.begin(), relTableNames.end());
```
Comment (andyfengHKU) [inline src/main/connection.cpp:233]: ditto

File: src/include/processor/operator/aggregate/aggregate_hash_table.h [hunk 8]
```diff
@@ -9,6 +9,13 @@
 namespace kuzu {
 namespace processor {
 
+template<typename type>
+bool compareEntryWithKeys(const uint8_t* keyValue, const uint8_t* entry) {
```
Comment (andyfengHKU) [inline src/include/processor/operator/aggregate/aggregate_hash_table.h:13]: ditto try put back. We have many places that are written in a similar way. I don't know why some of them can be compiled while others we need to change

File: src/include/processor/operator/order_by/order_by_key_encoder.h [hunk 9]
```diff
@@ -138,5 +133,9 @@ class OrderByKeyEncoder {
     std::vector<encode_function_t> encodeFunctions;
 };
 
+template<typename type>
```
Comment (andyfengHKU) [inline src/include/processor/operator/order_by/order_by_key_encoder.h:136]: ditto

File: src/include/function/vector_operations.h [hunk 10]
```diff
@@ -47,48 +47,46 @@ struct VectorOperationDefinition : public FunctionDefinition {
...
-
-    template<typename RESULT_TYPE, typename FUNC>
-    static void ConstExecFunction(const std::vector<std::shared_ptr<common::ValueVector>>& params,
-        common::ValueVector& result) {
-        assert(params.size() == 0);
-        ConstOperationExecutor::execute<RESULT_TYPE, FUNC>(result);
-    }
-};
+
+// Note: Previously part of a VectorOperations class, but MSVC is unable to resolve references to the function within subclasses.
```
Comment (andyfengHKU) [inline src/include/function/vector_operations.h:51]: I would still try to wrap them in a class. It's not a good idea to expose static function directly.

File: src/include/function/comparison/vector_comparison_operations.h (hunk 2 above)
Comment (benjaminwinger) [reply src/include/function/comparison/vector_comparison_operations.h:170]: My mistake; something left in from debugging. I'll remove it.

File: test/demo_db/demo_db_test.cpp (hunk 5 above)
Comment (benjaminwinger) [reply test/demo_db/demo_db_test.cpp:100]: I'm not sure, that was something @acquamarin did in [this commit](https://github.com/kuzudb/kuzu/pull/1573/commits/f35d4af43dff6420a254183777bbd628cc3c04da). @acquamarin?

File: src/main/connection.cpp (hunk 6 above)
Comment (benjaminwinger) [reply src/main/connection.cpp:219]: @acquamarin?

File: src/include/processor/operator/aggregate/aggregate_hash_table.h (hunk 8 above)
Comment (benjaminwinger) [reply src/include/processor/operator/aggregate/aggregate_hash_table.h:13]: Could you give an example of another place which didn't require changing? That might provide some insight into other ways the issue can be worked around.

It's worth noting that it seems to work fine as long as the function pointer isn't returned from another function (that's the point where the function resolution fails).

E.g. [this](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjEEgCspAAOqAqETgwe3r7%2ByanpAiFhkSwxcVyJdpgOGUIETMQEWT5%2BAVU1AnUNBEUR0bEJtvWNzTltwz2hfaUDFQCUtqhexMjsHAD06wDUACoAnkmYW3tLxFtoWFsIsZikWyRbtKhM6FuGW5iqrEn0AHQmGgAguYAMyhZDeS4mEFuKheBgdZi0aHYAHAoEQpgKBS7TAKAhbEwAdisQKSXiitDwyBAaK29K2BEwLB%2BTCZ0LcBAOjFYR0BAH0dgBNZSokGooEMrb4tnUrYAN1QeFeTPx/LhCMcAggAuFooVYi8mDmhJJdKlDPN9OJABE0bboaT0YDFcrGXiCBATcSnVL8egQCANYiOa7XhBQgQ5iiDd4jtCbYSzAA2HYewOqgjq%2BEhmGRlGOq2xo0QLjRkFO20cBa0TjxXh%2BDhaUioThuazWaWnVZJkE8UgETTVhYAa38iVrHEkDaHLc4vAUIA0A6HCzgsCQaBZdFi5EoW6SO7i8uQSSS/PlXAAnPyDJn%2Bapk9IsPLqZgAGp4TAAdwA8odG37GhaCZYhFwgKJZyiUIGj2Th%2B2g5hiD2X8om0apB24XgtzYQRfwYWg4KbXgsCiLxgDcMRaEXLDSCwFhDGAcRiLovBiAwvB5TxWdPmqLwmXg3hI0wSdmypKJiFgjwsFnAhiDwFhBIWKgDGABRPx/f9GEEmRBBEMR2CkXT5CUNRZ10Lh9EYlAO0sfQ8CiRdIAWVAki1BgaN4VAuOIeSsCcr1bBEjjnAgVxRj8SzgimEoyj0FI0nciL4vydzeli2YgvsdyuhGTwWj0docomdL%2BnKIZumSyyZUaUqZnKBYFG7QyazrGcWNbDgtkfSQthYBQTwVa9fjvD0tggXBCAeUEy14TCtDmUdx30Thp1IRtm06hclxXYi1xgRAQCWAhyQIPcIAPI9wl5ThVAADmTABaJ8tmAZBkC2LgQV%2BftMHwIg/L0fg9NEcQjOBkyVHUFiLNIb9JKSJSVo4et1tnTrf34077iobqXv6wbLyvEa2TGiAPG3egzhmuY5tXJazEkX4rxZ1m2bZ5NkbWjavPnWwdvm6tSHXQ7Lqp86xYGYAvssl83w0v8AJ04DQPAyCWMQ2CdM15DUPQhwdJwxgCHwwjZ1I8jKNoaidPoxjmObfB2JqLiaObXjkH4tZ%2B2E0TeHEyTkOktZmzkhSkZUpg1IVrTAN4CH9LB6QIcUKHzNyAwjBsyxrHsxz4BctyMk8lsfL87jnKykK/DChh3HynIorruq4sshKCkyBvIryRKMhbzKitqCYqqrxFcsmYoysK4eu%2Bn7p%2B4axZllWCRWpR9rNs4bqHue3q3o%2Br6fvGyaAd7WbdoWhmJ1WjfeY4bbl0FxbkbMW%2B53vi/h1IHy0mcSQgA%3D%3D) works, but [this](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjEEtIADqgKhE4MHt6%2B/qRJKY4CIWGRLDFxUraY9vkMQgRMxAQZPn4BdpgOabX1BIUR0bHxtnUNTVmtwz2hfSUDUgCUtqhexMjsHAD06wDUACoAngmYW3tLxFtoWFsIsZikWyRbtKhM6FuGW5iqrAn0AHQmGgAguYAMyhZDeS4mEFuKheBgdARiaHYAHAoEQpgKBS7TAKAhbEwAdisQISXiitDwyBAaK29K2BEwLB%2BTCZ0LcBAOjFYR0BAH0dgBNZSokGooEMrb4tnUrYAN1QeFeTPx/LhCOqEAFwtFCrEXkwc0JJLpUoZZvpxIAImibdDSejAfj0CAQBrEQwOYrlVsIKECHMUYy8QQIMbiY6pcRMARlgxCWYAGw7UNu1UEdXwz0cgMoh12onWjgLWicACsvD8HC0pFQnDc1ms0tOq0TIJ4pAImhLCwA1v5y/pOJIqz265xeAoQBouz2FnBYEg0Cy6LFyJQVwk13F5cgEgl%2BfKuABOfkGDP81RJ6RYeXUzAANTwmAA7gB5Q7Vzs0WhM4jThAUTjlEoT1HsnCdqBzDEHs75RNo7TdtwvArmwgjvgwtAQTWvBYFEXjAG4Yi0NOKGkFgLCGMA4i4RReAxh08p4uOnztF4TKQbwAaVOOVJRMQ4EeFg44EMQeAsFxCxUAYwAKM%2Bb6fowXEyIIIhiOw5T8IIigqOodG6Fw%2BjUSgTaWPoeBRNOkALKgCTVGRvCoMxxDiVg1nhhUVRpC4DDuJ4zR6MEUzFKUei5KkAijH4RkRdUvShbMXlIZ0EzRXobSel0DQJf0ZRDN06VGTKOUhXlEgLAoraaaWFZjnR9YcFs16SFsLAKHuCqnr8F6hn6uCEA8oJcHMvDIVocz9oOw4cKOpDVrWjVTjOc64QuMCICASwEOSBAbhAW47uEvKcKoAAcSYALQ3lswDIMgWxcCCvydpg%2BBEG5ejacIojiFpsi6Wo46GaQr6CQkUkzZW83jo174cbt9xUM1N3tZ1x4nj1bJ9RAHirvQZzDaNq0TVNZiSL8J5U9TNM00mM1zQtTmTrYK3jSWpCLpth0E/tPMDMAT1GXeD4KR%2BX4qb%2B/6AcBdHQeBKny7B8GIQ4KloYwBCYdh474YRxG0KRKmUdRtG1vgjGOMxZG1mxyAcWsnY8WWdH8YJsHCWstZiRJkMyUwcli0p368N96l/dI32A/pta6AEBhGKZljWBZVnwLZ9lpI5dYuW5LE2clnq%2Bf5mQxYEfm5TM%2BVxWkRU5MkkUMJXYXFZUKUCNljQBVkrfeR3EzN0lJVd6XGUD2VVcVYsyyrFPLvQ0zE5NedV03XdD1PS9/XvUNZgdsT7OTaQA5mEOLuM7DLPTrOh%2B1RwZj1YtLO36QLkpM4khAA) does not.

Comment (andyfengHKU) [reply src/include/processor/operator/aggregate/aggregate_hash_table.h:13]: I see. This is a good point. Then I wonder if we also pass functor reference rather than return a functor. 

For example in `vector_boolean_operation` we can do
```
void VectorBooleanOperations::bindExecFunction(
    ExpressionType expressionType, const binder::expression_vector& children, scalar_exec_func& func) {
    if (isExpressionBinary(expressionType)) {
        func = bindBinaryExecFunction(expressionType, children);
    } else {
        assert(isExpressionUnary(expressionType));
        func = bindUnaryExecFunction(expressionType, children);
    }
}
```
I think @acquamarin and @ray6080 should also comment on this. If so, @benjaminwinger you can ignore my other comment and do this in a separate PR.

File: src/storage/store/node_table.cpp [hunk 11]
```diff
@@ -26,6 +26,13 @@ void NodeTable::initializeData(NodeTableSchema* nodeTableSchema) {
     }
 }
 
+void NodeTable::resetColumns(catalog::NodeTableSchema* nodeTableSchema) {
```
Comment (ray6080) [inline src/storage/store/node_table.cpp:29]: Why do we need this ?

File: src/transaction/transaction_manager.cpp [hunk 12]
```diff
@@ -91,6 +91,7 @@ void TransactionManager::stopNewTransactionsAndWaitUntilAllReadTransactionsLeave
             numTimesWaited++;
             if (numTimesWaited * THREAD_SLEEP_TIME_WHEN_WAITING_IN_MICROS >
                 checkPointWaitTimeoutForTransactionsToLeaveInMicros) {
+                mtxForStartingNewTransactions.unlock();
```
Comment (ray6080) [inline src/transaction/transaction_manager.cpp:94]: Was this a bug?

File: src/storage/store/rel_table.cpp [hunk 13]
```diff
@@ -73,6 +73,21 @@ void DirectedRelTableData::initializeLists(RelTableSchema* tableSchema, WAL* wal
     }
 }
 
+void DirectedRelTableData::resetColumnsAndLists(
```
Comment (ray6080) [inline src/storage/store/rel_table.cpp:76]: Ditto here. Is this introduced in this PR?

File: src/main/database.cpp [hunk 14]
```diff
@@ -19,10 +25,19 @@ SystemConfig::SystemConfig() : SystemConfig(-1u) {}
 
 SystemConfig::SystemConfig(uint64_t bufferPoolSize_) {
     if (bufferPoolSize_ == -1u) {
+#if defined(_WIN32)
+        MEMORYSTATUSEX status;
+        status.dwLength = sizeof(status);
+        GlobalMemoryStatusEx(&status);
+        auto systemMemSize = (std::uint64_t)status.ullTotalPhys;
+        bufferPoolSize = (uint64_t)(BufferPoolConstants::DEFAULT_PHY_MEM_SIZE_RATIO_FOR_BM *
```
Comment (ray6080) [inline src/main/database.cpp:33]: Can we move this line `bufferPoolSize = (uint64_t)...` out of the `#if #else `?

**Review by ray6080 — COMMENTED** [review]
I have some comments. But I kind of want to take a look after `rebase` with master. As for now, it is not very clear some changes are already in master or made in this PR.

File: src/common/file_utils.cpp [hunk 15]
```diff
@@ -4,15 +4,64 @@
...
+        return -1;
+    }
+    return s.st_size;
+#endif
+}
+
 std::unique_ptr<FileInfo> FileUtils::openFile(const std::string& path, int flags) {
+#if defined(_WIN32)
+    // Not providing GENERIC_READ seems to cause problems.
+    int dwDesiredAccess = GENERIC_READ;
```
Comment (acquamarin) [inline src/common/file_utils.cpp:42]: Our convension is to use auto instead of the actual type

File: src/common/file_utils.cpp [hunk 16]
```diff
@@ -4,15 +4,64 @@
...
+    int dwCreationDisposition = OPEN_EXISTING;
+    int dwShareMode = FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE;
+    int dwFlagsAndAttributes = FILE_ATTRIBUTE_NORMAL;
+    if (flags & O_CREAT)
+        dwCreationDisposition = OPEN_ALWAYS;
```
Comment (acquamarin) [inline src/common/file_utils.cpp:47]: I think you can rewrite this as `auto dwCreationDisposition = (flags & O_CREAT) ?  OPEN_ALWAYS : OPEN_EXISTING`

File: src/common/file_utils.cpp [hunk 17]
```diff
@@ -4,15 +4,64 @@
...
+    if (flags & (O_CREAT | O_WRONLY | O_RDWR))
```
Comment (acquamarin) [inline src/common/file_utils.cpp:48]: Our convension is to always add brackets to if statements

File: src/common/file_utils.cpp [hunk 18]
```diff
@@ -125,5 +211,32 @@ std::vector<std::string> FileUtils::globFilePath(const std::string& path) {
     return result;
 }
 
+void FileUtils::truncateFileToSize(FileInfo* fileInfo, uint64_t size) {
+#if defined(_WIN32)
+    long offsetHigh = size >> 32;
```
Comment (acquamarin) [inline src/common/file_utils.cpp:216]: I think the size of long is dependent on the platform. Better to use `auto offsetHigh = (uint32_t)(size >> 32)`?

File: src/main/connection.cpp (hunk 6 above)
Comment (acquamarin) [reply src/main/connection.cpp:219]: We don't preserve the order of properties inside the catalog, we used unordered_map to store the nodeTableSchemas

File: src/storage/store/node_table.cpp (hunk 11 above)
Comment (acquamarin) [reply src/storage/store/node_table.cpp:29]: This function is used to close the files for nodeTable, if we don't close the files, we can't do the rename

File: src/common/file_utils.cpp (hunk 15 above)
Comment (benjaminwinger) [reply src/common/file_utils.cpp:42]: That doesn't seem to be mentioned in the [style guide](https://github.com/kuzudb/kuzu/wiki/GraphflowDB-Coding-Style). Is the style guide out of date?

File: external/arrow/apache_arrow.cmake [hunk 19]
```diff
@@ -2,13 +2,19 @@ include(ExternalProject)
 
 set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 
+if(WIN32)
```
[the remark this answers is not in the document]
Comment (benjaminwinger) [reply external/arrow/apache_arrow.cmake:5]: Sure, though it would trigger an unnecessary rebuild of arrow when switching on other platforms, and building arrow isn't very fast.

File: src/common/file_utils.cpp (hunk 18 above)
Comment (benjaminwinger) [reply src/common/file_utils.cpp:216]: It is platform-specific code though. I was following [the API for SetFilePointer](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-setfilepointer), which says the arguments are a `LONG`, which according to [this](https://learn.microsoft.com/en-us/windows/win32/winprog/windows-data-types) is just `typedef long LONG;`.

But I'll still change it; it's unintuitive.

Comment (benjaminwinger) [reply src/common/file_utils.cpp:216]: Though that actually would require casting the pointer back to a `long*`/`LONG*` when passing the argument to `SetFilePointer`, which seems a little redundant. Maybe it's better to just actually use `LONG` to indicate it's a microsoft-specified type for a microsoft-specified interface.

File: src/common/file_utils.cpp [hunk 20]
```diff
@@ -4,15 +4,62 @@
...
+#else
+    if (fd != -1) {
+        close(fd);
+    }
+#endif
+}
+
+int64_t FileInfo::getFileSize() {
+#ifdef _WIN32
+    return GetFileSize((HANDLE)handle, nullptr);
```
Comment (ray6080) [inline src/common/file_utils.cpp:29]: Should we use `GetFileSizeEx`? `GetFileSize` seems only return 4 bytes as the result, which is not large enough.

File: src/common/file_utils.cpp [hunk 21]
```diff
@@ -4,15 +4,62 @@
...
+    if (fstat(fd, &s) == -1) {
+        return -1;
+    }
+    return s.st_size;
+#endif
+}
+
 std::unique_ptr<FileInfo> FileUtils::openFile(const std::string& path, int flags) {
+#if defined(_WIN32)
+    // Not providing GENERIC_READ seems to cause problems.
```
Comment (ray6080) [inline src/common/file_utils.cpp:41]: Why is this? Can you take a look at this implementation? https://github.com/duckdb/duckdb/blob/09486d2b9d93270aa05920c6e7e54af00ce9d296/src/common/local_file_system.cpp#L524

File: src/include/processor/operator/aggregate/aggregate_hash_table.h (hunk 8 above)
Comment (ray6080) [reply src/include/processor/operator/aggregate/aggregate_hash_table.h:13]: > I see. This is a good point. Then I wonder if we also pass functor reference rather than return a functor. Because expose static function in header files is not a good coding style
> 
> For example in `vector_boolean_operation` we can do
> 
> ```
> void VectorBooleanOperations::bindExecFunction(
>     ExpressionType expressionType, const binder::expression_vector& children, scalar_exec_func& func) {
>     if (isExpressionBinary(expressionType)) {
>         func = bindBinaryExecFunction(expressionType, children);
>     } else {
>         assert(isExpressionUnary(expressionType));
>         func = bindUnaryExecFunction(expressionType, children);
>     }
> }
> ```
> 
> I think @acquamarin, @ray6080 and @benjaminwinger should all comment on this. If so, @benjaminwinger you can ignore my other comment and do this in a separate PR.

Sounds good to me.

**Review by andyfengHKU — APPROVED** [review]
